src/pudl_archiver/archivers/eia/eiarecs.py

```diff
@@ -0,0 +1,231 @@
+"""Archive EIA Residential Energy Consumption Survey (RECS)."""
+
+import logging
+import re
+from dataclasses import dataclass
+from io import BytesIO
+from urllib.parse import urljoin
+
+from pudl_archiver.archivers.classes import (
+    AbstractDatasetArchiver,
+    ArchiveAwaitable,
+    ResourceInfo,
+)
+from pudl_archiver.frictionless import ZipLayout
+
+logger = logging.getLogger(f"catalystcoop.{__name__}")
+
+
+@dataclass
+class LinkSet:
+    """Information a set of links in one tab of the RECS viewer.
+
+    See https://www.eia.gov/consumption/residential/data/2020/.
+    """
+
+    url: str
+    short_name: str
+    extension: str
+    pattern: re.Pattern
+    skip_if_html: bool = True
+
+
+def _url_for(year: int, view: str):
+    """Get the URL for a specific RECS year/tab combo."""
+    return (
+        f"https://www.eia.gov/consumption/residential/data/{year}/index.php?view={view}"
+    )
+
+
+# Each year, each tab's format changes. Rather than have complicated regexes that capture everything, just have lots of simple regexes
+YEAR_LINK_SETS = {
+    2020: {
+        "housing_characteristics": LinkSet(
+            url=_url_for(year=2020, view="characteristics"),
+            short_name="hc",
+            pattern=re.compile(r"HC (\d{1,2}\.\d{1,2})\.xlsx"),
+            extension="xlsx",
+        ),
+        "consumption & expenditures": LinkSet(
+            url=_url_for(year=2020, view="consumption"),
+            short_name="ce",
+            pattern=re.compile(r"ce(\d\.\d{1,2}[a-z]?)\.xlsx"),
+            extension="xlsx",
+        ),
+        "state data (housing characteristics)": LinkSet(
+            url=_url_for(year=2020, view="state"),
+            short_name="state-hc",
+            pattern=re.compile(r"State (.*)\.xlsx"),
+            extension="xlsx",
+        ),
+        "state data (consumption & expenditures)": LinkSet(
+            url=_url_for(year=2020, view="state"),
+            short_name="state-ce",
+            pattern=re.compile(r"ce(\d\.\d{1,2}\..*)\.xlsx"),
+            extension="xlsx",
+        ),
+        "microdata": LinkSet(
+            url=_url_for(year=2020, view="microdata"),
+            short_name="microdata",
+            pattern=re.compile(r"(recs.*public.*)\.csv"),
+            extension="csv",
+        ),
+        "microdata-codebook": LinkSet(
+            url=_url_for(year=2020, view="microdata"),
+            short_name="microdata",
+            pattern=re.compile(r"(RECS 2020 Codebook.*v.)\.xlsx"),
+            extension="xlsx",
+        ),
```

Comment:
not blocking but i would suggest making these a little more generic so there aren't so many to maintain

they both have the `year` and they both have `recs` if your add `re.IGNORECASE` and you could `(\.csv|\.xlsx)` for different file types. this is kinda a high level overall suggestion to condense patterns when possible but not a hill i will die on.